File: gui_apps/PyDoodle/step04/drawWindow.py
"""
// ============================================================================
// drawWindow.py: custom QMainWindow derived class for main window
//
// Tutorial - PyQt5 Doodle Application
// Based on a similar tutorial for Borland ObjectWindows Library (OWL)
// @author: Manish Bhobe
// My experiments with the Qt Framework. Use at your own risk!!
// ============================================================================
"""
import sys
from PyQt5.QtCore import *
from PyQt5.QtGui import *
from PyQt5.QtWidgets import *

class DrawWindow(QMainWindow):
    def __init__(self, *args, **kwargs):
        super(QMainWindow, self).__init__(*args, **kwargs)
        self.setWindowTitle("PyQt5 Doodle - Step04: Drawing a Squiggle")
        self.setStyleSheet("background-color: white")
        self.setGeometry(QRect(100,100,640,480))
        self.modified = False
        self.points = []
        self.dragging = False
        self.penColor = QColor(qRgb(0, 85, 255))
        self.penWidth = 3
        self.setMouseTracking(True)

    # operating system Events
    def closeEvent(self, e):
        if self.modified:
            resp = QMessageBox.question(self, "Confirm Close",
                                        "This will close the application.\nOk to quit?",
                                        QMessageBox.Yes | QMessageBox.No, QMessageBox.No)
            if resp == QMessageBox.Yes:
                e.accept()
            else:
                e.ignore()
        else:
            e.accept()

    def drawSquiggle(self, painter):
        if len(self.points) > 0:
            #print(f"In drawLine() function - drawing points {self.points}")
            pen = QPen(self.penColor, self.penWidth)
            painter.setPen(pen)
            lastPt = None
            for i, pt in enumerate(self.points):
                if i > 0:
                    painter.drawLine(lastPt, pt)
                lastPt = pt

    def paintEvent(self, e: QPaintEvent) -> None:
        painter = QPainter()
        try:
            painter.begin(self)
            painter.setRenderHint(QPainter.Antialiasing, True)
            self.drawSquiggle(painter)
        finally:
            painter.end()

    def mousePressEvent(self, e: QMouseEvent) -> None:
        if e.button() == Qt.LeftButton:
            # clear the previous squiggle
            self.points = []
            # start a new squiggle
            pt = QPoint(e.pos().x(), e.pos().y())
            #print(f"Got mousePressEvent at ({pt.x()}, {pt.y()})")
            qDebug(f"Got a Left-Mouse-Button-Pressed event at ({pt.x()}, {pt.y()})")
            self.points.append(pt)
            self.modified = True
            self.dragging = True
        elif e.button() == Qt.RightButton:
            qDebug("Got a Right-Mouse-Button-Pressed event")
            self.points = []
            self.modified = False
            self.update()

    def mouseMoveEvent(self, e: QMouseEvent) -> None:
        if (e.buttons() == Qt.LeftButton) and (self.dragging):
            pt = QPoint(e.pos().x(), e.pos().y())
            #print(f"Got mouseMoveEvent event at ({pt.x()}, {pt.y()})")
            self.points.append(pt)
            self.update()
        else:
            e.accept()

    def mouseReleaseEvent(self, e: QMouseEvent) -> None:
        if (e.button() == Qt.LeftButton) and (self.dragging):
            pt = QPoint(e.pos().x(), e.pos().y())
            self.points.append(pt)
            #print(f"Got mouseReleaseEvent event at ({pt.x()}, {pt.y()})")
            self.dragging = False
            self.update()
        else:
            e.accept()

File: gui_apps/Doodle/step06/DrawWindow.cc
// DrawWindow.cc: implements DrawWindow class
#include <QtGui>
#include <QApplication>
#include <QMessageBox>
#include <QInputDialog>
#include <QColorDialog>
#include "DrawWindow.h"
#include "Doodle.h"
#include "Line.h"
#include "common.hxx"

const QString AppTitle("Qt Scribble");
const QString WindowTitle("Qt Doodle - Step06: Drawing multiple lines");

DrawWindow::DrawWindow()
{
  setAttribute(Qt::WA_StaticContents);
  setWindowTitle(WindowTitle);
  resize(640,480);
  _dragging = false;
  _doodle = new Doodle();
  _currLine = 0;
}

DrawWindow::~DrawWindow()
{
  delete _doodle;
}

void DrawWindow::closeEvent(QCloseEvent *event)
{
  // window is about to close, prompt user & decide if ok to quit
  // based on user's response.
  if (_doodle->modified()) {
    switch(QMessageBox::question(this, tr("Qt Scribble Tutorial"),
          tr("This will close the application.\nOk to quit now?"),
          QMessageBox::Yes|QMessageBox::No, QMessageBox::No))
    {
      case QMessageBox::Yes:
        // ok to quit
        event->accept();
        break;
      default:
        // don't quit yet
        event->ignore();
    }
  }
}

void DrawWindow::drawLineTo(const QPoint& pt)
{
  Q_ASSERT(_currLine != nullptr);
  // draw line from _lastPt to pt
  QPainter painter(&_image);
  QPen pen(_currLine->penColor(), _currLine->penWidth());
  painter.setRenderHint(QPainter::Antialiasing);
  painter.setPen(pen);
  painter.drawLine(_lastPt, pt);
  _lastPt = pt;
  update();
}

void DrawWindow::clearImage()
{
  Q_ASSERT(_doodle != nullptr);
  _doodle->clear();
  QColor color = getPaletteColor(QPalette::Window);
  _image.fill(color);
  //_image.fill(qRgb(255,255,255));
  update();
}

void DrawWindow::mousePressEvent(QMouseEvent *event)
{
  Q_ASSERT(_doodle != nullptr);

  // check if Ctrl key is held down as left/right mouse is clicked
  Qt::KeyboardModifiers modifiers  = QApplication::queryKeyboardModifiers ();
  bool ctrlKeyIsDown = modifiers.testFlag( Qt::ControlModifier);

  qDebug() << "mousePressEvent() - CTRL key "
    << (ctrlKeyIsDown ? "IS" : "IS **NOT**")
    << " held down!";

  if (event->button() == Qt::LeftButton) {
    // left mouse button pressed
    if (ctrlKeyIsDown)
      changePenWidth();
    else {
      _currLine = _doodle->newLine();
      _lastPt = event->pos();
      _currLine->addPoint(_lastPt);
      _dragging = true;
      _doodle->setModified(true);
    }
  }
  else if (event->button() == Qt::RightButton) {
    if (ctrlKeyIsDown)
      changePenColor();
    else {
      clearImage();
      _doodle->setModified(false);
    }
  }
}

void  DrawWindow::mouseMoveEvent(QMouseEvent *event)
{
  if (event->buttons() & Qt::LeftButton && _dragging) {
    drawLineTo(event->pos());
    _currLine->addPoint(event->pos());
  }
}

void DrawWindow::mouseReleaseEvent(QMouseEvent *event)
{
  if ((event->button() == Qt::LeftButton) && _dragging) {
    drawLineTo(event->pos());
    _currLine->addPoint(event->pos());
    _dragging = false;
  }
}

void DrawWindow::changePenWidth()
{
  Q_ASSERT(_doodle != nullptr);
  // display message box & get width of pen
  bool ok;
  int newPenWidth = QInputDialog::getInt(this, AppTitle,
      QString("Enter new pen width:"), _doodle->penWidth(), 2, 10, 1, &ok);
  if (ok) {
    qDebug() << "New pen width selected: " << newPenWidth;
    _doodle->setPenWidth(newPenWidth);
  }
}

void DrawWindow::changePenColor()
{
  Q_ASSERT(_doodle != nullptr);
  // display standard color dialog & get new pen color
  QColor color = QColorDialog::getColor(_doodle->penColor(), this);
  if (color.isValid())
    _doodle->setPenColor(color);
}

void DrawWindow::resizeEvent(QResizeEvent *event)
{
  if (width() > _image.width() || height() > _image.height()) {
    int newWidth = qMax(width(), _image.width());
    int newHeight = qMax(height(), _image.height());
    resizeImage(QSize(newWidth, newHeight));
    update();
  }
  QWidget::resizeEvent(event);
}

void DrawWindow::paintEvent(QPaintEvent *event)
{
  QPainter painter(this);
  painter.setRenderHint(QPainter::Antialiasing);
  // we just blit the from image to device
  QRect dirtyRect = event->rect();
  painter.drawImage(dirtyRect, _image, dirtyRect);
}

void DrawWindow::resizeImage(const QSize& newSize)
{
  if (_image.size() == newSize)
    return;
  QImage newImage(newSize, QImage::Format_RGB32);
  QColor color = getPaletteColor(QPalette::Window);
  newImage.fill(color);
  //newImage.fill(qRgb(255,255,255));
  // draw existing image over new image
  QPainter painter(&newImage);
  painter.setRenderHint(QPainter::Antialiasing);
  painter.drawImage(QPoint(0,0), _image);
  _image = newImage;
}





